proposals/0000-custom-executors.md

```diff
@@ -0,0 +1,464 @@
+support hooking the default concurrent executor, perhaps by
+overriding a weak symbol, the same way that C programs can
+override `malloc`.  We won't discuss this further in this proposal
+because it doesn't otherwise impact the language and library design.
+
+### The default serial executor implementation
+
+The default serial executor implementation is separately instantiated
+for each actor that doesn't declare a custom executor; see "Actor
+executors" below. It is based on an "asynchronous lock" design which
+allows existing threads to immediately begin executing code on behalf
+of the actor rather than requiring functions to suspend and resume
+executing asynchronously. This process is called *switching*. In
+situations where switching is impossible, such as when the actor is
+already executing on a thread, a job to process the actor will be
+scheduled onto the default concurrent executor.
+
+Switching significantly complicates the design of serial executors,
+which can be seen in the `SerialExecutor` protocol below. This
+complexity could be eliminated by only allowing default serial
+executors to participate in switching. Currently, we are reluctant
+to hardcode this limitation into the protocol design.
+
+### Actor executors
+
+An actor's executor must conform to `SerialExecutor`, which refines
+the `Executor` protocol. `SerialExecutor` makes additional guarantees
+about the behavior of `execute(currentExecutor:)` and also adds
+several new methods which are used to allow the executor to opt in
+to supporting "switching" serial executors the same way that the
+default serial executor does.
+
+TODO: if we need to support blocking actors, we will need to extend
+this significantly; otherwise, actor executors will need to assume
+that returning from a job should always unblock the executor.
+(Presumably we do not want to block actor executors by blocking
+their thread.)
+
+```swift
+protocol SerialExecutor: Executor {
+  /// Is it possible for this executor to give up the current thread
+  /// and allow it to start running a different actor?
+  var canGiveUpThread: Bool { get }
+
+  /// Given that canGiveUpThread() previously returned true, give up
+  /// the current thread.
+  func giveUpThread()
+
+  /// Attempt to start running a task on the current actor.  Returns
+  /// true if this succeeds.
+  func tryClaimThread() -> Bool
+}
+```
+
+All actors implicitly conform to the `Actor` protocol.  An actor
+must provide a serial executor.
+
+```swift
+protocol Actor: AnyObject {
+  /// Return the serial executoor for this actor.
+  ///
+  /// This must always return the same reference, and `isSerial()`
+  /// must return true for the reference.
+  ///
+  /// Keeping the actor reference valid must be sufficient to keep
+  /// the executor reference valid.
+  var serialExecutor: UnownedExecutorRef { get }
+}
+```
+
+An `actor` may derive its executor implementation in one
+of the following ways. We may add more ways in the future.
```

Comment:
These seem a bit brittle to me: if you have a typo in `serialExecutor`, for example, presumably the actor would use the default serial executor and you would be none-the-wiser?

Reply:
I see your point, but I'm not sure what to do about it.  It's not reasonable to require actors to explicitly opt in to using the default serial executor, which means that there's always going to be some series of errors of omission that can lead to using it accidentally.

If we had a way to annotate that a function is supposed to be used for a protocol — a long-standing request — that would work for the `serialExecutor` case specifically.

Reply:
I'm not too worried here to be honest. Same issue as with any other customization hook like that...

If you REALLY wanted to make sure to not make that mistake I think there's a number of ways to address it in plain user code.

a) @glbrntt I think you could make a require the executor type by a protocol:

```

protocol Executor {}
class EventLoop: Executor {}
protocol Actor {
  associatedtype ActorExecutor = Executor
  var ex: ActorExecutor { get }
}
protocol NeedsEL: Actor {
    associatedtype ActorExecutor = EventLoop
}

class Bad: NeedsEL {
  // error: repl.swift:10:7: error: type 'Specific' does not conform to protocol 'Actor'
  // class Specific: NeedsEL {
  //      ^
  //
  // repl.swift:5:7: note: protocol requires property 'ex' with type 'EventLoop'; do you want to add a stub?
  //  var ex: ActorExecutor { get }
}

class Good: NeedsEL {
  let ex: EventLoop
  init(executor eventLoop: EventLoop) {
    self.ex = eventLoop
  }
}
```

This should be possible right? Synthesis would have to be aware or it would fail since the default serial executor would not fulfil this requirement.

Reply:
> It's not reasonable to require actors to explicitly opt in to using the default serial executor, which means that there's always going to be some series of errors of omission that can lead to using it accidentally.

I agree. 

I'm also not worried about this, but seemed worth pointing out as a potential pitfall. Calling it out in documentation would probably be sufficient.